assembly/src/parsers/io_ops.rs

```diff
@@ -255,70 +270,104 @@ pub fn parse_loc_storew(op: &Token) -> Result<Node, ParsingError> {
+        .collect()
 }
 
-fn parse_hex_params(
-    op: &Token,
-    param_idx: usize,
+fn parse_push_one(
+    token: &Token,
+    param_type: ParamType,
     param_str: &str,
-) -> Result<Vec<Felt>, ParsingError> {
-    // handle error cases where the hex string is poorly formed
-    let is_single_element = if param_str.len() <= HEX_CHUNK_SIZE {
-        if param_str.len() % 2 != 0 {
-            // parameter string is not a valid hex representation
-            return Err(ParsingError::invalid_param(op, param_idx));
-        }
-        true
+) -> Result<Node, ParsingError> {
+    let value = if param_type == ParamType::Decimal {
+        parse_decimal_param(token, 1, param_str)?
+    } else if param_type == ParamType::Hex {
+        parse_hex_param(token, 1, param_str)?
     } else {
-        if param_str.len() % HEX_CHUNK_SIZE != 0 {
-            // hex string doesn't contain a valid number of bytes
-            return Err(ParsingError::invalid_param(op, param_idx));
-        }
-        false
+        // If we find invalid ParamType, we have internal parsing error and should panic
+        panic!("Invalid parameter type")
     };
+    if value <= u8::MAX as u64 {
+        Ok(Instruction(PushU8(value as u8)))
+    } else if value <= u16::MAX as u64 {
+        Ok(Instruction(PushU16(value as u16)))
+    } else if value <= u32::MAX as u64 {
+        Ok(Instruction(PushU32(value as u32)))
+    } else if value < Felt::MODULUS {
+        Ok(Instruction(PushFelt(get_valid_felt(token, 1, value)?)))
+    } else {
+        Err(ParsingError::invalid_param(token, 1))
+    }
+}
 
-    // parse the hex string into one or more valid field elements
-    if is_single_element {
-        // parse a single element in hex representation
-        let parsed_param = parse_hex_param(op, param_idx, param_str)?;
-        Ok(vec![parsed_param])
+fn parse_push_many(
+    token: &Token,
+    param_type: ParamType,
+    param_str: &str,
+) -> Result<Node, ParsingError> {
+    let values = if param_type == ParamType::Hex {
+        parse_hex_params(token, param_str)?
+    } else if param_type == ParamType::HexAndDecimal {
+        parse_constants(token)?
     } else {
-        // iterate over the multi-value hex string and parse each 8-byte chunk into a valid element
-        (0..param_str.len())
-            .step_by(HEX_CHUNK_SIZE)
-            .map(|i| parse_hex_param(op, param_idx, &param_str[i..i + HEX_CHUNK_SIZE]))
-            .collect()
+        // If we find invalid ParamType, we have internal parsing error and should panic
+        panic!("Invalid parameter type")
+    };
+    let values_max = values.iter().max();
+    match values_max {
+        Some(value) if *value <= u8::MAX as u64 => Ok(Instruction(PushU8List(
+            values.iter().map(|v| *v as u8).collect(),
+        ))),
+        Some(value) if *value <= u16::MAX as u64 => Ok(Instruction(PushU16List(
+            values.iter().map(|v| *v as u16).collect(),
+        ))),
```

Comment:
3 nested expressions in a single line is quite difficult to read. Consider expanding this code into some `let` expressions